torchmultimodal/modules/layers/position_embedding.py

```diff
@@ -169,3 +170,106 @@ def forward(self, t: Tensor) -> Tensor:
+        then in forward of transformer layer:
+            alibi_mask = self.alibi.get_attention_mask(N) # N = seq length of this batch
+            ...
+            attn = q @ k.transpose( -2, -1)
+            att *= 1.0 / math.sqrt(k.size(-1))
+            att += alibi_mask
+
+        """
+        super().__init__()
+
+        self.num_heads = num_heads
+        self.max_seq_len = max_seq_len
+
+        self.causal_mask = self.build_causal_attention_mask(
+            self.max_seq_len, self.num_heads
+        )
+        self.alibi_mask_base = self.build_alibi_mask(self.max_seq_len, self.num_heads)
+        self.decoder_mask = self.causal_mask + self.alibi_mask_base
+        self.register_buffer("alibi_mask", self.decoder_mask, persistent=False)
+
+    def get_attention_mask(self, curr_seq_len: int) -> torch.Tensor:
+        """returns the alibi mask, clipped to the current batch seq len"""
+        return self.alibi_mask[..., :curr_seq_len, :curr_seq_len]
+
+    @classmethod
+    def build_causal_attention_mask(cls, seq_len: int, num_heads: int) -> torch.Tensor:
+        """builds a generic causal attention mask"""
+        causal_mask = torch.triu(
+            torch.ones(seq_len, seq_len) * float("-inf"), diagonal=1
+        )
+        attn_mask = causal_mask.repeat(num_heads, 1, 1)
+        return attn_mask
+
+    @classmethod
+    def build_alibi_mask(cls, seq_len: int, num_heads: int) -> torch.Tensor:
+        """generate the alibi mask by computing a distance bias matrix multiplied by each head's m (slope)"""
+        distance_bias_matrix = -torch.abs(
+            torch.arange(seq_len) - torch.arange(seq_len).view(-1, 1)
+        )
+        slope_per_head = Tensor(cls.get_slopes(num_heads)).view(-1, 1, 1)
+        alibi_mask = distance_bias_matrix * slope_per_head
+        return alibi_mask
+
+    @staticmethod
+    def get_slopes(num_heads: int) -> List[float]:
+        """for n heads, a range from (0,1) and is the geometric sequence
+        that starts at 2^(-8/n) and uses this same value as its ratio
+
+        example: num_heads =4
+        result: [0.25, 0.0625, 0.015625, 0.00390625]
+
+        """
+
+        def get_slopes_power_of_2(n: int) -> List[float]:
+            start = 2 ** (-(2 ** -(math.log2(n) - 3)))
+            ratio = start
+            return [start * ratio**i for i in range(n)]
+
+        if math.log2(num_heads).is_integer():
+            return get_slopes_power_of_2(num_heads)
+
+        # paper authors note that they only trained models that have 2^a heads for some a.
+        # This has beneficial properties related to input being power of 2.
+        # Closest power of 2 below is workaround for when num of heads is not power of 2
+
+        closest_power_of_2 = 2 ** math.floor(math.log2(num_heads))
+        return (
+            get_slopes_power_of_2(closest_power_of_2)
+            + get_slopes_power_of_2(2 * closest_power_of_2)[0::2][
+                : num_heads - closest_power_of_2
```

Comment:
Looks good, my one strictly cosmetic gripe is that this returns slopes in a fairly unintuitive order for non-powers of 2: for 12 heads, for example, the scales are `2^(1,2,3,4,5,6,7,8,.5,1.5,2.5,3.5)`. Rather than doing:
```
a = get_slopes_power_of_2(closest_power_of_2)
b = get_slopes_power_of_2(2 * closest_power_of_2)[0::2][:num_heads - closest_power_of_2]
return a+b
```
You could return something like:
```
return [x for pair in zip(b,a) for x in pair] + a[len(b):]
```
which, while less legible, would return scales in order (i.e. for 12 heads this gives `2^(.5,1,1.5,2,2.5,3,3.5,4,5,6,7,8)`). 

Up to you if you think that tradeoff is worth it.

Reply:
Thanks for the feedback @daviswer  - I agree, I think the consistency and symmetry of the heads (regardless of power of 2) has merit and will update.

Reply:
I've updated as well as added unit test coverage for non power of 2 heads case. 
Thanks again!